### ISSUE TITLE
Classes in Shoes

### ISSUE THREAD

**Issue by dredknight** [shoes/shoes3#385]
Just started using ruby classes for as it looks as a very good habit that I do not have. Unfortunately I bumped into issue where Shoes commands cannot be used in the class. I recall @ccoupe telling me about that a long time ago but I do not remember what exactly was the problem.

How can I go around the problem?

Here is a simple example.
```Ruby
class My_sign
     def initialize(text)
           @my_text = text
     end
      def in_front_of_the_door_says
          para @my_text
      end
end

Shoes.app do
       wooden_sign = My_sign.new("beware of the dog")
       wooden_sign.in_front_of_the_door_says
end
```

**Comment by BackOrder** [comment]
Numinoes illustrates perfectly how to mix traditional software development approach (including unit testing) and Shoes GUI programming. Basically, you can pass a reference to Shoes.app object to your class and then use reflection to send method calls.

```ruby
class My_sign
   def initialize(shoes, text)
      @shoes = shoes
      @text = text
   end

   def in_front_of_the_door_says
      @shoes.send(:para, @text)
   end
end

Shoes.app do
   wooden_sign = My_sign.new(self, "beware of the dog")
   wooden_sign.in_front_of_the_door_says
end
```

**Comment by dredknight** [comment]
Thanks! What is being a pain in the back(self pointing always to shoes.app ) can also be helpful :).

**Comment by BackOrder** [comment]
Ain't that so? aka #26 :)

**Comment by BackOrder** [comment]
The good news is that you can retain most of the Shoes feeling even with reflection. The sliding out effect in Numinoes uses reflection and sends animate to Shoes. Notice that it uses a block (`{}`, `do ... end` would also work) as one would in Shoes. Hope this helps.

![image](https://user-images.githubusercontent.com/8844578/32662247-de93b036-c5f7-11e7-8233-e5dd2df78f92.png)

![out](https://user-images.githubusercontent.com/8844578/32662209-b9adc5c2-c5f7-11e7-8fb1-8020562e389b.gif)

**Comment by dredknight** [comment]
Thank you  ^_^. That animated menu is awesome!

**Comment by dredknight** [comment]
@BackOrder  there was a way to make a widget class. Something like this:
```Ruby
class do_something << Shoes::Widget
      make my own stuff
end
```

As far as I remember this make your class inherit all the things Widget class has.
Where I can see what widget does and how is this different from the example you gave above with the .send command.

**Comment by BackOrder** [comment]
Shoes widgets never really took off for some reason. HacketyHack extensively uses it but most Shoes app just don't. I remember @ccoupe said something about it that it ain't working so well. It would be a good thing to explore and expand on the topic.

The code I gave you above is to be able to communicate with a Shoes.app object **outside** of Shoes DSL, that is, in plain Ruby. It allowed me to create a sort of [MVC](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) for which it is possible to run unit tests on models and made it easier to write reusable UI code through classes/methods. Shoes DSL is a rather descriptive language, reusable is not its strong point.

Basically, those two are different things. The method presented here is for advanced ~ the topic would be **Using reflection in Shoes**. You asked how to do it and so I explained to you but you would be correct to believe that your code snippet may be more appropriate for the use of a Shoes widget.

https://github.com/shoes/shoes3/wiki/Widgets-and-Canvas
https://github.com/shoes/shoes3/blob/master/samples/expert/tooltips.rb

**Comment by ccoupe** [comment]
> I remember @ccoupe said something about it that it ain't working so well. It would be a good thing to explore and expand on the topic.

I was trying to implement a tree  widget/tableview - a music browser, player like RhythmBox or Itunes . Beyond my skill level.

**Comment by BackOrder** [comment]
> I was trying to implement a tree widget/tableview - a music browser, player like RhythmBox or Itunes . Beyond my skill level.

That would be really cool. By the way, what is wrong with GtkTreeView and NSOutlineView/NSTreeControl?

**References**
https://developer.gnome.org/gtk3/stable/GtkTreeView.html
https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/OutlineView/OutlineView.html
https://developer.apple.com/documentation/appkit/nsoutlineview
https://developer.apple.com/documentation/appkit/nstreecontroller
